File: sword_offer/python/27_字符串的排列.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

import time
import itertools

class Solution:
    def __init__(self):
        self.ret = []

    # ===== 外部接口 =====
    def Permutation1(self, ss):
        """
        1. 使用stl求解（有序，无重复）
        注意：
            [1] stl中的全排列函数，存在重复，且不排序
            [2] 返回值是一个迭代器，使用list()后，内部元素都是tuple
        """
        if len(ss) == 0:
            return []
        
        tmp_ss = sorted(ss)
        tmp_ret = ["".join(i) for i in itertools.permutations(tmp_ss)]

        # 利用set去重并排序
        tmp_ret = list(set(tmp_ret))
        tmp_ret.sort()
        return tmp_ret
    
    def Permutation2(self, ss):
        """
        2. 递归传递引用，set，sort（有序，无重复）
        """
        if len(ss) == 0:
            return []
        
        # 无序版本
        self.perm0(list(ss), 0)

        # 利用set去重并排序
        self.ret = list(set(self.ret))
        self.ret.sort()
        return self.ret
    
    def Permutation3(self, ss):
        """
        3. 递归传递引用，内部使用stl排序（有序，重复）
        """
        if len(ss) == 0:
            return []
        
        self.perm1(sorted(list(ss)), 0)
        return self.ret

    def Permutation4(self, ss):
        """
        4. 递归传递值，不使用使用stl排序，循环替换实现（有序，重复）
        """
        if len(ss) == 0:
            return []
        
        self.perm2(sorted(list(ss)), 0)
        return self.ret

    # ===== 递归函数 =====

    def perm0(self, strList, start):
        """
        无序版本，递归传递引用
        输入字符串ss对应的列表strList，以及起始索引start
        原始字符串无法快速交换元素值
        """
        if start == len(strList) - 1:
            self.ret.append("".join(strList))
            return
        
        for i in range(start, len(strList)):
            strList[i], strList[start] = strList[start], strList[i]
            self.perm0(strList, start + 1)
            strList[i], strList[start] = strList[start], strList[i]
        return
    
    def perm1(self, strList, start):
        """
        有序版本，递归传递引用
        输入字符串ss对应的列表strList，以及起始索引start
        注意：
            [1] 每次递归调用sort排序
            [2] 如果不要求有序，可以省去sort()步骤
            [3] 需要额外的步骤，来检测重复性
        """
        if start == len(strList) - 1:
            self.ret.append("".join(strList))
            return
        
        # 为了防止排序会打乱索引，导致上一层调用无法恢复
        # 这里保存排序之前的str
        oriStr = strList[:]
        strList = strList[0:start] + sorted(strList[start:])
        
        for i in range(start, len(strList)):
            if i == start or (strList[i] not in strList[start:i]):
                strList[i], strList[start] = strList[start], strList[i]
                self.perm1(strList, start + 1)
                strList[i], strList[start] = strList[start], strList[i]
        
        strList = oriStr
        return

    def perm2(self, oriStrList, start):
        """
        有序版本，递归传递值
        输入字符串ss对应的列表strList，以及起始索引start
        注意：
            [1] 这种方式传递了很多次形式参数
            [2] 不需要额外的步骤
        """
        if start == len(oriStrList) - 1:
            self.ret.append("".join(oriStrList))
            return
        
        # 保证值传递
        strList = oriStrList[:]
        
        for i in range(start, len(strList)):
            if i == start or strList[i] != strList[start]:
                strList[i], strList[start] = strList[start], strList[i]
                self.perm1(strList, start + 1)
        
        return
        
    # ===== 工具函数 =====
    
    def printf_vecStr(self, ret):
        """
        打印list[str]结果
        仅显示结果个数不超过24的情况
        """
        if len(ret) > 24:
            print("{:d} numbers, too many to display!".format(len(ret)))
        else:
            print(", ".join(ret))


    def test(self, ss):
        """
        测试函数
        """
        func_vec = [self.Permutation1,
                    self.Permutation2,
                    self.Permutation3,
                    self.Permutation4]
        print("=====")
        for func in func_vec:
            self.ret = []
            tmp_ss = ss[:]

            start = time.time()
            result = func(tmp_ss)
            end = time.time()

            print("tims(us): {:>7.2f}, result: ".format((end - start)*10**6), end="")
            self.printf_vecStr(result)


def main():
    str1 = "cab"
    str2 = "abac"
    str3 = "cbba"
    str4 = "sghi"
    str5 = "iugoa"
    str6 = "sadjbg"

    s = Solution()
    s.test(str1)
    s.test(str2)
    s.test(str3)
    s.test(str4)
    s.test(str5) #结果个数：5! = 120
    s.test(str6) # 结果个数：6! = 720


if __name__ == "__main__":
    main()

File: sword_offer/cpp/27_字符串的排列.cpp
#include <iostream>
#include <vector>
#include <set>
#include <map>
#include <algorithm>
#include <stdio.h>
#include <sys/time.h>

class Solution {
public:
    std::vector<std::string> ret;

    // ===== 外部接口 =====

    // 1. 使用stl求解（有序，无重复）
    std::vector<std::string> Permutation1(std::string str) {
        if (str.empty()) {
            return ret;
        }
        
        sort(str.begin(), str.end());

        do {
            ret.push_back(str);
        } while (std::next_permutation(str.begin(), str.end()));

        return ret;
    }

    // 2. 递归传递引用，set,sort（有序，无重复）
    std::vector<std::string> Permutation2(std::string str) {
        if (str.empty()) {
            return ret;
        }
        
        // 无序版本
        perm0(str, 0);

        // 利用set对vec去重复
        std::set<std::string> set(ret.begin(), ret.end());
        ret.assign(set.begin(), set.end());

        // 利用stl对vec排序
        sort(ret.begin(), ret.end());
        return ret;
    }

    // 3. 递归传递引用，内部使用stl排序（有序，无重复）
    std::vector<std::string> Permutation3(std::string str) {
        if (str.empty()) {
            return ret;
        }
        
        sort(str.begin(), str.end());
        perm1(str, 0); // 调用perm-method-1
        return ret;
    }

    // 4. 递归传递值，不使用stl排序，循环替换实现（有序，无重复）
    std::vector<std::string> Permutation4(std::string str) {
        if (str.empty()) {
            return ret;
        }
        
        sort(str.begin(), str.end());
        perm2(str, 0); // 调用perm-method-2
        return ret;
    }

    // ===== 递归函数 =====

    // 无序版本
    void perm0(std::string& str, int start) {
        if (start == str.size() - 1) {
            ret.push_back(str);
            return;
        }

        for (int i = start; i < str.size(); i++) {
            swap(str[i], str[start]);
            perm0(str, start + 1);
            swap(str[i], str[start]);
        }
        return;
    }


    // 说明：
    //     [1] 递归求解，将当前的全排列添加到ret
    //     [2] 输入str引用，起始索引
    // 存在的问题：
    //     [1] 每次递归调用sort排序，使用引用
    //     [2] 如果不要求有序，可以省去sort()步骤
    //     [3] 需要调用额外的函数，检测重复性
    void perm1(std::string& str, int start) {
        if (start == str.size() - 1) {
            ret.push_back(str);
            return;
        }
        
        // 注意在传递引用时内部排序，后打乱str，使用上一层swap无法恢复str，导致出现重复
        // 所以这里保存排序之前的str
        std::string oriStr = str;
        sort(str.begin() + start, str.end());
        
        // 依次交换，并进行下一次的迭代
        for (int i = start; i < str.size(); i++) {
            if (i == start || (str[start] != str[i] && !ifHasSame(str, start, i))) {
                swap(str[i], str[start]); // 交换
                perm1(str, start + 1);    // 递归
                swap(str[i], str[start]); // 交换回来
            }
        }

        str = oriStr; // 保证排序不影响上一层递归恢复
        return;
    }

    // ===== 其它方式：优化版 =====
    // 改动：
    //     [1] 递归内部传值，而非引用
    //     [2] 去除了不必要的形式参数
    //     [3] 不使用sort，借助循环交换来实现有序
    // 缺点：
    //     [1] 这种方式传递了很多次形式参数
    //     [2] 不需要额外的防重复逻辑
    void perm2(std::string str, int start) {
        if (start == str.size() - 1) {
            ret.push_back(str);
            return;
        }
        
        // 依次交换，并进行下一次的迭代
        for (int i = start; i < str.size(); i++) {
            // 跳过重复
            if (i == start || str[start] != str[i]) {
                swap(str[i], str[start]); // 交换
                perm2(str, start + 1); // 递归
            }
        }
        return;
    }

    // ===== 工具函数 =====
    // 交换string中两个字符的值
    void swap(char& a, char& b) {
        char tmp = a;
        a = b;
        b = tmp;
    }

    // 判断[start, i]之间是否有与i相同的元素
    bool ifHasSame(std::string& str, int start, int i) {
        for (int k = start + 1; k < i; k++) {
            if (str[k] == str[i]) {
                return true;
            }
        }
        return false;
    }

    // 打印结果，仅显示结果个数不超过24的情况
    void printf_vecStr(std::vector<std::string>& vec) {
        if (vec.size() > 24) {
            printf("%ld numbers, too many to display!\n", vec.size());
            return;
        }

        for (auto i : vec) {
            std::cout << i << ", ";
        }
        std::cout << std::endl;
        return;
    }

    // 测试函数
    void test(std::string& str) {
        std::vector<std::string> result;
        struct timeval start, end;
        printf("=====\n");
        for (auto func : this->func_vec_) {
            this->ret.clear();
            gettimeofday(&start, nullptr);
            result = (this->*func)(str);
            gettimeofday(&end, nullptr);
            printf("time(us): %4ld, result: ", end.tv_usec - start.tv_usec);
            this->printf_vecStr(result);
        }
    }

private:
    typedef std::vector<std::string> (Solution::*func_ptr)(std::string);
    std::vector<func_ptr> func_vec_ = {&Solution::Permutation1,
                                       &Solution::Permutation2,
                                       &Solution::Permutation3,
                                       &Solution::Permutation4};
};

int main(int argc, char* argv[])
{
    std::string str1 = "cab";
    std::string str2 = "abac";
    std::string str3 = "cbba";
    std::string str4 = "sghi";
    std::string str5 = "iugoa";
    std::string str6 = "sadjbg";

    Solution s;
    s.test(str1);
    s.test(str2);
    s.test(str3);
    s.test(str4);
    s.test(str5); // 结果个数：5! = 120
    s.test(str6); // 结果个数：6! = 720

    return 0;
}